src/jit/lower.cpp

```diff
@@ -1881,251 +1880,315 @@ GenTree* Lowering::LowerTailCallViaHelper(GenTreeCall* call, GenTree* callTarget
+            // approaches that require flow control. However, the generated code is slightly
+            // larger.
+            //
 
-        genTreeOps hiCmpOper;
-        genTreeOps loCmpOper;
+            GenTree* src1 = cmp->gtGetOp1();
+            GenTree* src2 = cmp->gtGetOp2();
 
-        switch (cmp->OperGet())
-        {
-            case GT_LT:
-                cmp->gtOper = GT_GT;
-                hiCmpOper   = GT_LT;
-                loCmpOper   = GT_LT;
-                break;
-            case GT_LE:
-                cmp->gtOper = GT_GT;
-                hiCmpOper   = GT_LT;
-                loCmpOper   = GT_LE;
-                break;
-            case GT_GT:
-                cmp->gtOper = GT_LT;
-                hiCmpOper   = GT_GT;
-                loCmpOper   = GT_GT;
-                break;
-            case GT_GE:
-                cmp->gtOper = GT_LT;
-                hiCmpOper   = GT_GT;
-                loCmpOper   = GT_GE;
-                break;
-            default:
-                unreached();
-        }
+            //
+            // Set the comparison type to TYP_BYTE, otherwise we'll get one MOVZX instruction for
+            // each comparison. We'll add a cast from TYP_UBYTE (yes, unsigned because we get a
+            // zero latency MOVZX instead of a MOVSX) after the select to compensate.
+            //
+            // Note that there's no byte-size CMOV instruction so we'll end up using a register
+            // after a partial update. That may cause a stall but this doesn't appear to degrade
+            // performance if the low 8 byte registers are used.
+            //
 
-        BasicBlock* newBlock2 = comp->fgSplitBlockAtEnd(newBlock);
+            cmp->gtType     = TYP_BYTE;
+            cmp->gtOp.gtOp1 = src1->gtGetOp1();
+            cmp->gtOp.gtOp2 = src2->gtGetOp1();
 
-        GenTree* hiJcc = new (comp, GT_JCC) GenTreeJumpCC(hiCmpOper);
-        hiJcc->gtFlags = cmp->gtFlags;
-        LIR::AsRange(newBlock).InsertAfter(nullptr, hiJcc);
+            GenTree* hiCmp = comp->gtNewOperNode(cmp->OperGet(), TYP_BYTE, src1->gtGetOp2(), src2->gtGetOp2());
 
-        BlockRange().Remove(loSrc1.Def());
-        BlockRange().Remove(loSrc2.Def());
-        GenTree* loCmp   = comp->gtNewOperNode(loCmpOper, TYP_INT, loSrc1.Def(), loSrc2.Def());
-        loCmp->gtFlags   = cmp->gtFlags | GTF_UNSIGNED;
-        GenTree* loJtrue = comp->gtNewOperNode(GT_JTRUE, TYP_VOID, loCmp);
-        LIR::AsRange(newBlock2).InsertAfter(nullptr, loSrc1.Def(), loSrc2.Def(), loCmp, loJtrue);
+            hiCmp->gtFlags = cmp->gtFlags;
+            cmp->gtFlags |= GTF_UNSIGNED;
 
-        m_block->bbJumpKind = BBJ_COND;
-        m_block->bbJumpDest = nextDest;
-        nextDest->bbFlags |= BBF_JMP_TARGET;
-        comp->fgAddRefPred(nextDest, m_block);
+            GenTree* select = new (comp, GT_SELCC) GenTreeSelectCC(GT_EQ, TYP_INT, hiCmp, cmp);
```

Comment:
For `EQ`/`NE` compares we could use `AND`/`OR` instead of `CMOV` and save a code byte. However, `AND`/`OR` don't handle well the `TYP_BYTE` trick and we end up with more code.